tests/benchmark/crud_test.go

```diff
@@ -0,0 +1,895 @@
+
+func readAndGZipFile(filename string) (*bytes.Buffer, error) {
+	fileBytes, _ := ioutil.ReadFile(filename)
+	var buf bytes.Buffer
+	zw, _ := gzip.NewWriterLevel(&buf, gzip.BestCompression)
+	_, err := zw.Write(fileBytes)
+	if err != nil {
+		return nil, err
+	}
+	if err := zw.Close(); err != nil {
+		return nil, err
+	}
+	return &buf, nil
+}
+
+func BenchmarkCreateApi(b *testing.B) {
+	ctx := context.Background()
+	gapicClient, err := connection.NewClient(ctx)
+	if err != nil {
+		b.Fatalf("Setup: Failed to create client: %s", err)
+	}
+	client := gapicClient.GrpcClient()
+
+	for i := 1; i <= b.N; i++ {
+		if i%1000 == 0 {
+			//Sleep 10s every 1000 iterations to reduce server load
+			time.Sleep(10 * time.Second)
+		}
+
+		apiId := getApiName(i)
+		req := &rpc.CreateApiRequest{
+			Parent: rootResource,
+			ApiId:  apiId,
+			Api: &rpc.Api{
+				DisplayName: apiId,
+				Description: fmt.Sprintf("Description for %s", apiId),
+			},
+		}
+		b.StartTimer()
+		_, err := client.CreateApi(ctx, req)
+		b.StopTimer()
+		/**
+		Go Benchmarking always runs all the benchmark tests once before it runs
+		for the duration specified. So the first row always returns error.
+		*/
+		if i > 1 && err != nil {
+			b.Errorf("CreateApi(%+v) returned unexpected error: %s", req, err)
+		}
+	}
+}
+
+func BenchmarkCreateApiVersion(b *testing.B) {
+	ctx := context.Background()
+	gapicClient, err := connection.NewClient(ctx)
+	if err != nil {
+		b.Fatalf("Setup: Failed to create client: %s", err)
+	}
+	client := gapicClient.GrpcClient()
+
+	b.StopTimer()
+	b.ResetTimer()
+
+	for i := 1; i <= b.N; i++ {
+		if i%1000 == 0 {
+			//Sleep 10s every 1000 iterations to reduce server load
+			time.Sleep(10 * time.Second)
+		}
+
+		apiId := getApiName(i)
+		for j := 1; j <= getNumberOfVersions(); j++ {
+			req := &rpc.CreateApiVersionRequest{
+				Parent: fmt.Sprintf("%s/apis/%s", rootResource, apiId),
+				ApiVersion: &rpc.ApiVersion{
+					DisplayName: fmt.Sprintf("v%d", j),
+					Description: fmt.Sprintf("v%d of %s", j, apiId),
+				},
+				ApiVersionId: fmt.Sprintf("v%d", j),
+			}
+			// Only take reading for the first version
+			if j == 1 {
+				b.StartTimer()
+			}
+			_, err := client.CreateApiVersion(ctx, req)
+			if j == 1 {
+				b.StopTimer()
+			}
+			if i > 1 && err != nil {
+				b.Errorf("CreateApiVersion(%+v) returned unexpected error: %s", req, err)
+			}
+		}
+	}
+}
+
+func BenchmarkCreateApiSpec(b *testing.B) {
+	ctx := context.Background()
+	gapicClient, err := connection.NewClient(ctx)
+	if err != nil {
+		b.Fatalf("Setup: Failed to create client: %s", err)
+	}
+	client := gapicClient.GrpcClient()
+
+	b.StopTimer()
+	b.ResetTimer()
+
+	for i := 1; i <= b.N; i++ {
+		if i%1000 == 0 {
+			//Sleep 10s every 1000 iterations to reduce server load
+			time.Sleep(10 * time.Second)
+		}
+
+		apiId := getApiName(i)
+
+		for j := 1; j <= getNumberOfVersions(); j++ {
+
+			buf, err := readAndGZipFile(filepath.Join("testdata", "openapi1.yaml"))
```

Comment:
This seems unnecessarily expensive. With some of our examples this file will be re-read from disk 10,000 times during the benchmark. Why not read it once at the start?

Reply:
done